### PR TITLE
feat: added dynamic tags suggestion and fixed search by tag

### LINKED ISSUE #115: [FEATURE 💡] Tag Sugessions
### Description

- Fetch all tags from the database using a method. 
- Then display tag suggestions to select from to the user using AutoComplete.
- This feature can be added in Forms so that users don't create duplicate tags and use existing ones.
- If you are unable to connect to database because u don't have mongoDB string please do it using some mock tags stored in an array in frontEnd itself.
- See this [page](https://mui.com/material-ui/react-autocomplete/) for the Autocomplete component

### Screenshots

image

download

### Additional information

- Always use styled components instead of plain CSS and MUI components.
## Happy contributing. 💐
Star my other Repositories [here](https://github.com/warmachine028)

### PR DESCRIPTION
**Pull request by JayeshYadav99**
# Issue #115  📐

- I [Jayesh ](https://github.com/JayeshYadav99) have worked for #115

This pull request introduces two main changes:

1. **Dynamic Tag Suggestion**: I have added a dynamic tag suggestion feature, which allows users to receive tag suggestions while performing a specific action (please specify when this action occurs).

2. **Controller for Getting All Tags**: I've added a controller to fetch all tags from the relevant source (please specify the source, e.g., a database or an API).

## Changes Made

- Added dynamic tag suggestion functionality.
- Implemented a controller for retrieving all tags.

## Guidelines 🔐

I accept the fact that I have followed the guidelines and have not copied the codes from around the internet

- [x] **Contribution Guidelines**
- [x] **Code of Conduct**

## Issue to be closed 🛅

- My pull request closes #115
## Screenshots 📷

Here are the pictures of changes that I have made 🔽

  
Please test it locally beacuse i have made changes on server also

  
---

### CODE REVIEW

File: client/.env.example
Comment (warmachine028) [inline client/.env.example:1]: Restore this file

File: client/src/components/Form/Form.jsx
Comment (warmachine028) [inline client/src/components/Form/Form.jsx:1]: Can you delete the ChipInput import commented out at the top

File: client/src/components/Search/Search.jsx [hunk 1]
```diff
@@ -39,7 +70,7 @@ const Search = ({ tags, setTags }) => {
 			<AppBar className={`${classes.searchBarLight} ${mode === 'light' ? classes.searchBarLight : classes.searchBarDark}`} position="static" color="inherit">
 				<TextField sx={{ input: { color: 'white' } }} name="search" variant="outlined" label="Search Memories" onKeyDown={handleKeyPress} fullWidth value={search} onChangeCapture={(e) => setSearch(e.target.value)} />
 				{/* <ChipInput label="Search Tags" InputProps={{ style: { color: 'white' } }} value={tags} newChipKeyCodes={[188, 13]} onAdd={handleAdd} onDelete={handleDelete} variant="outlined" className={classes.chip} /> */}
```
Comment (warmachine028) [inline client/src/components/Search/Search.jsx:72]: Please don't remove this comment by mistake.

File: server/.env.example
Comment (warmachine028) [inline server/.env.example:1]: Restore this file

File: server/controllers/posts.js [hunk 2]
```diff
@@ -102,6 +102,7 @@ export const getPost = async (req, res) => {
 
 export const getPostsBySearch = async (req, res) => {
 	const { searchQuery, tags } = req.query
+	console.log(tags)
 	try {
```
Comment (warmachine028) [inline server/controllers/posts.js:106]: Remove this console.log

File: server/controllers/posts.js [hunk 3]
```diff
@@ -111,10 +112,14 @@ export const getPostsBySearch = async (req, res) => {
...
 				{
-					$or: [{ title }, { tags: { $in: tags.split(',') } }],
+					$and: [
+						{ title }, // Include title condition if not empty
+						tags ? { tags: { $in: tags.split(',') } } : {}, // Include tags condition if tags are given
+					  ],
 				},
 			],
 		}
+		console.log(query);
```
Comment (warmachine028) [inline server/controllers/posts.js:122]: Remove this console.log

File: server/controllers/posts.js [hunk 4]
```diff
@@ -111,10 +112,14 @@ export const getPostsBySearch = async (req, res) => {
 					$or: [{ creator: userId }, { private: false }],
 				},
 				{
-					$or: [{ title }, { tags: { $in: tags.split(',') } }],
+					$and: [
```
Comment (warmachine028) [inline server/controllers/posts.js:115]: Explain the changes u made in this query

File: server/controllers/posts.js [hunk 5]
```diff
@@ -128,6 +133,7 @@ export const getPostsBySearch = async (req, res) => {
 			},
 		])
 		posts = setCreator(posts)
+		console.log(posts)
```
Comment (warmachine028) [inline server/controllers/posts.js:136]: Remove this console.log

**Review by warmachine028 — CHANGES_REQUESTED** [review]

File: server/controllers/posts.js (hunk 4 above)
Comment (JayeshYadav99) [reply server/controllers/posts.js:115]: I have made this changes beacuse your tags search was giving all posts on searching specific tags  and empty title ,with and it is working fine

**Comment by JayeshYadav99** [comment]
@warmachine028 ,I have made changes please review

**Comment by warmachine028** [comment]
Change the Title

**Comment by JayeshYadav99** [comment]
> Change the Title

what should be title  can you tell me?

**Comment by JayeshYadav99** [comment]
@warmachine028 ,done

**Comment by warmachine028** [comment]
Hello Contributor,
- Make sure to see all the Closed PRs for the proper PR title and PR template
- Read this issue for the [reason]( https://github.com/warmachine028/nike/issues/151) for delay in PR approvals

**Comment by JayeshYadav99** [comment]
@warmachine028 ,any problem?

**Comment by warmachine028** [comment]
> @warmachine028 ,any problem?

No your PR title is fine. I am right now unable to review PRs. That's why I stated the reason in the comment above.

**Pull request merged**